File: packages/less/src/less/tree/property.test.js
import { expect } from 'vitest';
import Property from './property';
import Declaration from './declaration';
import Value from './value';
import Anonymous from './anonymous';

describe('Property', () => {
    describe('constructor', () => {
        it('should initialize with name, index and fileInfo', () => {
            const name = 'test-property';
            const index = 42;
            const fileInfo = { filename: 'test.less' };
            const property = new Property(name, index, fileInfo);

            expect(property.name).toBe(name);
            expect(property._index).toBe(index);
            expect(property._fileInfo).toBe(fileInfo);
        });

        it('should handle empty name', () => {
            const property = new Property('', 0, { filename: 'test.less' });
            expect(property.name).toBe('');
        });

        it('should handle negative index', () => {
            const property = new Property('test', -1, {
                filename: 'test.less'
            });
            expect(property._index).toBe(-1);
        });

        it('should handle undefined fileInfo', () => {
            const property = new Property('test', 0, undefined);
            expect(property._fileInfo).toBeUndefined();
        });

        it('should handle null name', () => {
            const property = new Property(null, 0, { filename: 'test.less' });
            expect(property.name).toBeNull();
        });

        it('should handle undefined index', () => {
            const property = new Property('test', undefined, {
                filename: 'test.less'
            });
            expect(property._index).toBeUndefined();
        });

        it('should handle null fileInfo', () => {
            const property = new Property('test', 0, null);
            expect(property._fileInfo).toBeNull();
        });
    });

    describe('eval', () => {
        const createContext = (frames = []) => ({
            frames,
            importantScope: [{}],
            pluginManager: {
                less: {
                    visitors: {
                        ToCSSVisitor: {
                            prototype: {
                                _mergeRules: (arr) => arr
                            }
                        }
                    }
                }
            }
        });

        it('should throw error for recursive property reference', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            property.evaluating = true;

            expect(() => property.eval(createContext())).toThrow(
                expect.objectContaining({
                    type: 'Name',
                    message: 'Recursive property reference for test',
                    filename: 'test.less',
                    index: 0
                })
            );
        });

        it('should throw error for undefined property', () => {
            const property = new Property('undefined-property', 0, {
                filename: 'test.less'
            });
            const context = createContext();

            expect(() => property.eval(context)).toThrow(
                expect.objectContaining({
                    type: 'Name',
                    message: "Property 'undefined-property' is undefined",
                    filename: 'test.less',
                    index: 0
                })
            );
        });

        it('should evaluate property from frames', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const value = new Value([new Anonymous('value')]);
            const declaration = new Declaration(
                'test',
                value,
                false,
                false,
                0,
                { filename: 'test.less' }
            );

            const frame = {
                property: (name) => {
                    if (name === 'test') {
                        return [declaration];
                    }
                    return null;
                }
            };

            const context = createContext([frame]);
            const result = property.eval(context);

            // The value is returned as a Node object
            expect(result).toEqual(
                expect.objectContaining({
                    value: 'value',
                    allowRoot: true,
                    rulesetLike: false
                })
            );
        });

        it('should handle important flag in property evaluation', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const value = new Value([new Anonymous('value')]);
            const declaration = new Declaration(
                'test',
                value,
                '!important',
                false,
                0,
                { filename: 'test.less' }
            );

            const frame = {
                property: (name) => {
                    if (name === 'test') {
                        return [declaration];
                    }
                    return null;
                }
            };

            const importantScope = { important: false };
            const context = createContext([frame]);
            context.importantScope = [importantScope];

            property.eval(context);
            expect(importantScope.important).toBe(' !important');
        });

        it('should handle multiple declarations with merge', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const value1 = new Value([new Anonymous('value1')]);
            const value2 = new Value([new Anonymous('value2')]);
            const declaration1 = new Declaration(
                'test',
                value1,
                false,
                true,
                0,
                { filename: 'test.less' }
            );
            const declaration2 = new Declaration(
                'test',
                value2,
                false,
                true,
                0,
                { filename: 'test.less' }
            );

            const frame = {
                property: (name) => {
                    if (name === 'test') {
                        return [declaration1, declaration2];
                    }
                    return null;
                }
            };

            const context = createContext([frame]);
            const result = property.eval(context);

            // The value is returned as a Node object
            expect(result).toEqual(
                expect.objectContaining({
                    value: 'value2',
                    allowRoot: true,
                    rulesetLike: false
                })
            );
        });

        it('should handle empty frames array', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const context = createContext([]);

            expect(() => property.eval(context)).toThrow(
                expect.objectContaining({
                    type: 'Name',
                    message: "Property 'test' is undefined"
                })
            );
        });

        it('should handle multiple frames with same property', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const value1 = new Value([new Anonymous('value1')]);
            const value2 = new Value([new Anonymous('value2')]);
            const declaration1 = new Declaration(
                'test',
                value1,
                false,
                false,
                0,
                { filename: 'test.less' }
            );
            const declaration2 = new Declaration(
                'test',
                value2,
                false,
                false,
                0,
                { filename: 'test.less' }
            );

            const frame1 = {
                property: (name) => (name === 'test' ? [declaration1] : null)
            };
            const frame2 = {
                property: (name) => (name === 'test' ? [declaration2] : null)
            };

            const context = createContext([frame1, frame2]);
            const result = property.eval(context);
            expect(result.value).toBe('value1'); // Should take first frame's value
        });

        it('should handle multiple frames with different properties', () => {
            const property = new Property('test2', 0, {
                filename: 'test.less'
            });
            const value1 = new Value([new Anonymous('value1')]);
            const value2 = new Value([new Anonymous('value2')]);
            const declaration1 = new Declaration(
                'test1',
                value1,
                false,
                false,
                0,
                { filename: 'test.less' }
            );
            const declaration2 = new Declaration(
                'test2',
                value2,
                false,
                false,
                0,
                { filename: 'test.less' }
            );

            const frame1 = {
                property: (name) => (name === 'test1' ? [declaration1] : null)
            };
            const frame2 = {
                property: (name) => (name === 'test2' ? [declaration2] : null)
            };

            const context = createContext([frame1, frame2]);
            const result = property.eval(context);
            expect(result.value).toBe('value2');
        });

        it('should handle single declaration with merge flag', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const value = new Value([new Anonymous('value')]);
            const declaration = new Declaration('test', value, false, true, 0, {
                filename: 'test.less'
            });

            const frame = {
                property: (name) => (name === 'test' ? [declaration] : null)
            };

            const context = createContext([frame]);
            const result = property.eval(context);
            expect(result.value).toBe('value');
        });

        it('should handle multiple declarations without merge', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const value1 = new Value([new Anonymous('value1')]);
            const value2 = new Value([new Anonymous('value2')]);
            const declaration1 = new Declaration(
                'test',
                value1,
                false,
                false,
                0,
                { filename: 'test.less' }
            );
            const declaration2 = new Declaration(
                'test',
                value2,
                false,
                false,
                0,
                { filename: 'test.less' }
            );

            const frame = {
                property: (name) =>
                    name === 'test' ? [declaration1, declaration2] : null
            };

            const context = createContext([frame]);
            const result = property.eval(context);
            expect(result.value).toBe('value2'); // Takes last declaration when merge is false
        });

        it('should handle property with null value', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const value = new Value([new Anonymous(null)]);
            const declaration = new Declaration(
                'test',
                value,
                false,
                false,
                0,
                { filename: 'test.less' }
            );

            const frame = {
                property: (name) => (name === 'test' ? [declaration] : null)
            };

            const context = createContext([frame]);
            const result = property.eval(context);
            expect(result.value).toBeNull();
        });

        it('should handle property with undefined value', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const value = new Value([new Anonymous(undefined)]);
            const declaration = new Declaration(
                'test',
                value,
                false,
                false,
                0,
                { filename: 'test.less' }
            );

            const frame = {
                property: (name) => (name === 'test' ? [declaration] : null)
            };

            const context = createContext([frame]);
            const result = property.eval(context);
            expect(result.value).toBeUndefined();
        });

        it('should handle property with empty string value', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const value = new Value([new Anonymous('')]);
            const declaration = new Declaration(
                'test',
                value,
                false,
                false,
                0,
                { filename: 'test.less' }
            );

            const frame = {
                property: (name) => (name === 'test' ? [declaration] : null)
            };

            const context = createContext([frame]);
            const result = property.eval(context);
            expect(result.value).toBe('');
        });

        it('should handle properties with mixed merge flags', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const value1 = new Value([new Anonymous('value1')]);
            const value2 = new Value([new Anonymous('value2')]);
            const declaration1 = new Declaration(
                'test',
                value1,
                false,
                true,
                0,
                { filename: 'test.less' }
            );
            const declaration2 = new Declaration(
                'test',
                value2,
                false,
                false,
                0,
                { filename: 'test.less' }
            );

            const frame = {
                property: (name) =>
                    name === 'test' ? [declaration1, declaration2] : null
            };

            const context = createContext([frame]);
            const result = property.eval(context);
            expect(result.value).toBe('value2');
        });

        it('should handle multiple important flags', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const value1 = new Value([new Anonymous('value1')]);
            const value2 = new Value([new Anonymous('value2')]);
            const declaration1 = new Declaration(
                'test',
                value1,
                '!important',
                false,
                0,
                { filename: 'test.less' }
            );
            const declaration2 = new Declaration(
                'test',
                value2,
                '!important',
                false,
                0,
                { filename: 'test.less' }
            );

            const frame = {
                property: (name) =>
                    name === 'test' ? [declaration1, declaration2] : null
            };

            const importantScope = { important: false };
            const context = createContext([frame]);
            context.importantScope = [importantScope];

            property.eval(context);
            expect(importantScope.important).toBe(' !important');
        });
    });

    describe('find', () => {
        it('should find first matching item in array', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const arr = [1, 2, 3, 4];
            let found = false;
            const result = property.find(arr, (item) => {
                if (item === 3) {
                    found = true;
                    return item;
                }
                return null;
            });
            expect(found).toBe(true);
            expect(result).toBe(3);
        });

        it('should return null if no match found', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const arr = [1, 2, 3, 4];
            const result = property.find(arr, (item) => item > 5);
            expect(result).toBeNull();
        });

        it('should return null for empty array', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const result = property.find([], () => true);
            expect(result).toBeNull();
        });

        it('should handle callback returning undefined', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const arr = [1, 2, 3];
            const result = property.find(arr, () => undefined);
            expect(result).toBeNull();
        });

        it('should handle array with null values', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const arr = [null, 2, null];
            const result = property.find(arr, (item) => {
                if (item === 2) {
                    return item;
                }
                return null;
            });
            expect(result).toBe(2);
        });

        it('should handle callback throwing error', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const arr = [1, 2, 3];
            expect(() => {
                property.find(arr, () => {
                    throw new Error('Test error');
                });
            }).toThrow('Test error');
        });

        it('should handle array with undefined values', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const arr = [undefined, 2, undefined];
            const result = property.find(arr, (item) => {
                if (item === 2) {
                    return item;
                }
                return null;
            });
            expect(result).toBe(2);
        });

        it('should handle array with mixed types', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const arr = [1, 'string', true, null];
            const result = property.find(arr, (item) => {
                if (typeof item === 'string') {
                    return item;
                }
                return null;
            });
            expect(result).toBe('string');
        });

        it('should handle array with objects', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const obj = { key: 'value' };
            const arr = [1, obj, 3];
            const result = property.find(arr, (item) => {
                if (typeof item === 'object' && item !== null) {
                    return item;
                }
                return null;
            });
            expect(result).toBe(obj);
        });

        it('should handle array with functions', () => {
            const property = new Property('test', 0, { filename: 'test.less' });
            const func = () => {};
            const arr = [1, func, 3];
            const result = property.find(arr, (item) => {
                if (typeof item === 'function') {
                    return item;
                }
                return null;
            });
            expect(result).toBe(func);
        });
    });
});

File: packages/less/src/less/go_parser/property_test.go
package go_parser

import (
	"testing"
)

func TestProperty(t *testing.T) {
	t.Run("constructor", func(t *testing.T) {
		t.Run("should initialize with name, index and fileInfo", func(t *testing.T) {
			name := "test-property"
			index := 42
			fileInfo := map[string]any{"filename": "test.less"}
			property := NewProperty(name, index, fileInfo)

			if property.name != name {
				t.Errorf("Expected name %s, got %s", name, property.name)
			}
			if property.GetIndex() != index {
				t.Errorf("Expected index %d, got %d", index, property.GetIndex())
			}
			if property.FileInfo()["filename"] != fileInfo["filename"] {
				t.Errorf("Expected fileInfo %v, got %v", fileInfo, property.FileInfo())
			}
		})

		t.Run("should handle empty name", func(t *testing.T) {
			property := NewProperty("", 0, map[string]any{"filename": "test.less"})
			if property.name != "" {
				t.Errorf("Expected empty name, got %s", property.name)
			}
		})

		t.Run("should handle negative index", func(t *testing.T) {
			property := NewProperty("test", -1, map[string]any{"filename": "test.less"})
			if property.GetIndex() != -1 {
				t.Errorf("Expected index -1, got %d", property.GetIndex())
			}
		})
	})

	t.Run("eval", func(t *testing.T) {
		createContext := func(frames []any) map[string]any {
			return map[string]any{
				"frames": frames,
				"importantScope": []any{map[string]any{}},
				"mergeRules": func(arr []any) {},
			}
		}

		t.Run("should throw error for recursive property reference", func(t *testing.T) {
			property := NewProperty("test", 0, map[string]any{"filename": "test.less"})
			property.evaluating = true

			_, err := property.Eval(createContext(nil))
			if err == nil {
				t.Error("Expected error for recursive property reference")
			}
		})

		t.Run("should throw error for undefined property", func(t *testing.T) {
			property := NewProperty("undefined-property", 0, map[string]any{"filename": "test.less"})
			context := createContext(nil)

			_, err := property.Eval(context)
			if err == nil {
				t.Error("Expected error for undefined property")
			}
		})

		t.Run("should evaluate property from frames", func(t *testing.T) {
			property := NewProperty("test", 0, map[string]any{"filename": "test.less"})
			value, err := NewValue([]any{NewAnonymous("value", 0, nil, false, false, nil)})
			if err != nil {
				t.Fatalf("Failed to create value: %v", err)
			}
			declaration, err := NewDeclaration(
				"test",
				value,
				false,
				false,
				0,
				map[string]any{"filename": "test.less"},
				false,
				false,
			)
			if err != nil {
				t.Fatalf("Failed to create declaration: %v", err)
			}

			frame := map[string]any{
				"property": func(name string) []any {
					if name == "test" {
						return []any{declaration}
					}
					return nil
				},
			}

			context := createContext([]any{frame})
			result, err := property.Eval(context)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result == nil {
				t.Error("Expected non-nil result")
			}
		})

		t.Run("should handle property with null value", func(t *testing.T) {
			property := NewProperty("test", 0, map[string]any{"filename": "test.less"})
			value, err := NewValue([]any{NewAnonymous(nil, 0, nil, false, false, nil)})
			if err != nil {
				t.Fatalf("Failed to create value: %v", err)
			}
			declaration, err := NewDeclaration(
				"test",
				value,
				false,
				false,
				0,
				map[string]any{"filename": "test.less"},
				false,
				false,
			)
			if err != nil {
				t.Fatalf("Failed to create declaration: %v", err)
			}

			frame := map[string]any{
				"property": func(name string) []any {
					if name == "test" {
						return []any{declaration}
					}
					return nil
				},
			}

			context := createContext([]any{frame})
			result, err := property.Eval(context)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result == nil {
				t.Error("Expected non-nil result")
			}
		})

		t.Run("should handle property with empty string value", func(t *testing.T) {
			property := NewProperty("test", 0, map[string]any{"filename": "test.less"})
			value, err := NewValue([]any{NewAnonymous("", 0, nil, false, false, nil)})
			if err != nil {
				t.Fatalf("Failed to create value: %v", err)
			}
			declaration, err := NewDeclaration(
				"test",
				value,
				false,
				false,
				0,
				map[string]any{"filename": "test.less"},
				false,
				false,
			)
			if err != nil {
				t.Fatalf("Failed to create declaration: %v", err)
			}

			frame := map[string]any{
				"property": func(name string) []any {
					if name == "test" {
						return []any{declaration}
					}
					return nil
				},
			}

			context := createContext([]any{frame})
			result, err := property.Eval(context)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result == nil {
				t.Error("Expected non-nil result")
			}
		})

		t.Run("should handle properties with mixed merge flags", func(t *testing.T) {
			property := NewProperty("test", 0, map[string]any{"filename": "test.less"})
			value1, err := NewValue([]any{NewAnonymous("value1", 0, nil, false, false, nil)})
			if err != nil {
				t.Fatalf("Failed to create value1: %v", err)
			}
			value2, err := NewValue([]any{NewAnonymous("value2", 0, nil, false, false, nil)})
			if err != nil {
				t.Fatalf("Failed to create value2: %v", err)
			}
			declaration1, err := NewDeclaration(
				"test",
				value1,
				false,
				true,
				0,
				map[string]any{"filename": "test.less"},
				false,
				false,
			)
			if err != nil {
				t.Fatalf("Failed to create declaration1: %v", err)
			}
			declaration2, err := NewDeclaration(
				"test",
				value2,
				false,
				false,
				0,
				map[string]any{"filename": "test.less"},
				false,
				false,
			)
			if err != nil {
				t.Fatalf("Failed to create declaration2: %v", err)
			}

			frame := map[string]any{
				"property": func(name string) []any {
					if name == "test" {
						return []any{declaration1, declaration2}
					}
					return nil
				},
			}

			context := createContext([]any{frame})
			result, err := property.Eval(context)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result == nil {
				t.Error("Expected non-nil result")
			}
		})

		t.Run("should handle property with undefined value", func(t *testing.T) {
			property := NewProperty("test", 0, map[string]any{"filename": "test.less"})
			value, err := NewValue([]any{NewAnonymous(nil, 0, nil, false, false, nil)})
			if err != nil {
				t.Fatalf("Failed to create value: %v", err)
			}
			declaration, err := NewDeclaration(
				"test",
				value,
				false,
				false,
				0,
				map[string]any{"filename": "test.less"},
				false,
				false,
			)
			if err != nil {
				t.Fatalf("Failed to create declaration: %v", err)
			}

			frame := map[string]any{
				"property": func(name string) []any {
					if name == "test" {
						return []any{declaration}
					}
					return nil
				},
			}

			context := createContext([]any{frame})
			result, err := property.Eval(context)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result == nil {
				t.Error("Expected non-nil result")
			}
		})

		t.Run("should handle multiple important flags", func(t *testing.T) {
			property := NewProperty("test", 0, map[string]any{"filename": "test.less"})
			value1, err := NewValue([]any{NewAnonymous("value1", 0, nil, false, false, nil)})
			if err != nil {
				t.Fatalf("Failed to create value1: %v", err)
			}
			value2, err := NewValue([]any{NewAnonymous("value2", 0, nil, false, false, nil)})
			if err != nil {
				t.Fatalf("Failed to create value2: %v", err)
			}
			declaration1, err := NewDeclaration(
				"test",
				value1,
				"!important",
				false,
				0,
				map[string]any{"filename": "test.less"},
				false,
				false,
			)
			if err != nil {
				t.Fatalf("Failed to create declaration1: %v", err)
			}
			declaration2, err := NewDeclaration(
				"test",
				value2,
				"!important",
				false,
				0,
				map[string]any{"filename": "test.less"},
				false,
				false,
			)
			if err != nil {
				t.Fatalf("Failed to create declaration2: %v", err)
			}

			frame := map[string]any{
				"property": func(name string) []any {
					if name == "test" {
						return []any{declaration1, declaration2}
					}
					return nil
				},
			}

			importantScope := map[string]any{"important": false}
			context := createContext([]any{frame})
			context["importantScope"] = []any{importantScope}

			_, err = property.Eval(context)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if importantScope["important"] != " !important" {
				t.Errorf("Expected important flag to be set, got %v", importantScope["important"])
			}
		})

		t.Run("should handle empty frames array", func(t *testing.T) {
			property := NewProperty("test", 0, map[string]any{"filename": "test.less"})
			context := createContext([]any{})

			_, err := property.Eval(context)
			if err == nil {
				t.Error("Expected error for undefined property")
			}
		})

		t.Run("should handle multiple frames with same property", func(t *testing.T) {
			property := NewProperty("test", 0, map[string]any{"filename": "test.less"})
			value1, err := NewValue([]any{NewAnonymous("value1", 0, nil, false, false, nil)})
			if err != nil {
				t.Fatalf("Failed to create value1: %v", err)
			}
			value2, err := NewValue([]any{NewAnonymous("value2", 0, nil, false, false, nil)})
			if err != nil {
				t.Fatalf("Failed to create value2: %v", err)
			}
			declaration1, err := NewDeclaration(
				"test",
				value1,
				false,
				false,
				0,
				map[string]any{"filename": "test.less"},
				false,
				false,
			)
			if err != nil {
				t.Fatalf("Failed to create declaration1: %v", err)
			}
			declaration2, err := NewDeclaration(
				"test",
				value2,
				false,
				false,
				0,
				map[string]any{"filename": "test.less"},
				false,
				false,
			)
			if err != nil {
				t.Fatalf("Failed to create declaration2: %v", err)
			}

			frame1 := map[string]any{
				"property": func(name string) []any {
					if name == "test" {
						return []any{declaration1}
					}
					return nil
				},
			}
			frame2 := map[string]any{
				"property": func(name string) []any {
					if name == "test" {
						return []any{declaration2}
					}
					return nil
				},
			}

			context := createContext([]any{frame1, frame2})
			result, err := property.Eval(context)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result == nil {
				t.Error("Expected non-nil result")
			}
		})

		t.Run("should handle multiple frames with different properties", func(t *testing.T) {
			property := NewProperty("test2", 0, map[string]any{"filename": "test.less"})
			value1, err := NewValue([]any{NewAnonymous("value1", 0, nil, false, false, nil)})
			if err != nil {
				t.Fatalf("Failed to create value1: %v", err)
			}
			value2, err := NewValue([]any{NewAnonymous("value2", 0, nil, false, false, nil)})
			if err != nil {
				t.Fatalf("Failed to create value2: %v", err)
			}
			declaration1, err := NewDeclaration(
				"test1",
				value1,
				false,
				false,
				0,
				map[string]any{"filename": "test.less"},
				false,
				false,
			)
			if err != nil {
				t.Fatalf("Failed to create declaration1: %v", err)
			}
			declaration2, err := NewDeclaration(
				"test2",
				value2,
				false,
				false,
				0,
				map[string]any{"filename": "test.less"},
				false,
				false,
			)
			if err != nil {
				t.Fatalf("Failed to create declaration2: %v", err)
			}

			frame1 := map[string]any{
				"property": func(name string) []any {
					if name == "test1" {
						return []any{declaration1}
					}
					return nil
				},
			}
			frame2 := map[string]any{
				"property": func(name string) []any {
					if name == "test2" {
						return []any{declaration2}
					}
					return nil
				},
			}

			context := createContext([]any{frame1, frame2})
			result, err := property.Eval(context)
			if err != nil {
				t.Errorf("Unexpected error: %v", err)
			}
			if result == nil {
				t.Error("Expected non-nil result")
			}
		})
	})

	t.Run("find", func(t *testing.T) {
		t.Run("should find first matching item in array", func(t *testing.T) {
			property := NewProperty("test", 0, map[string]any{"filename": "test.less"})
			arr := []any{1, 2, 3, 4}
			found := false
			result := property.Find(arr, func(item any) any {
				if item == 3 {
					found = true
					return item
				}
				return nil
			})
			if !found {
				t.Error("Expected to find item")
			}
			if result != 3 {
				t.Errorf("Expected result 3, got %v", result)
			}
		})

		t.Run("should return nil if no match found", func(t *testing.T) {
			property := NewProperty("test", 0, map[string]any{"filename": "test.less"})
			arr := []any{1, 2, 3, 4}
			result := property.Find(arr, func(item any) any {
				if item.(int) > 5 {
					return item
				}
				return nil
			})
			if result != nil {
				t.Errorf("Expected nil result, got %v", result)
			}
		})

		t.Run("should handle array with null values", func(t *testing.T) {
			property := NewProperty("test", 0, map[string]any{"filename": "test.less"})
			arr := []any{nil, 2, nil}
			result := property.Find(arr, func(item any) any {
				if item == 2 {
					return item
				}
				return nil
			})
			if result != 2 {
				t.Errorf("Expected result 2, got %v", result)
			}
		})

		t.Run("should handle array with mixed types", func(t *testing.T) {
			property := NewProperty("test", 0, map[string]any{"filename": "test.less"})
			arr := []any{1, "string", true, nil}
			result := property.Find(arr, func(item any) any {
				if str, ok := item.(string); ok {
					return str
				}
				return nil
			})
			if result != "string" {
				t.Errorf("Expected result 'string', got %v", result)
			}
		})

		t.Run("should handle array with objects", func(t *testing.T) {
			property := NewProperty("test", 0, map[string]any{"filename": "test.less"})
			obj := map[string]any{"key": "value"}
			arr := []any{1, obj, 3}
			result := property.Find(arr, func(item any) any {
				if m, ok := item.(map[string]any); ok {
					return m
				}
				return nil
			})
			if result == nil {
				t.Error("Expected non-nil result")
			}
			if resultMap, ok := result.(map[string]any); !ok {
				t.Error("Expected result to be a map")
			} else if resultMap["key"] != obj["key"] {
				t.Errorf("Expected result map to have key 'value', got %v", resultMap["key"])
			}
		})
	})
} 